### PR TITLE
fix: added rcCommand[COLLECTIVE] to blackbox logging

### PR DESCRIPTION
**Pull request by egonl**
Added collective by refactoring RF1 for RF2. Now you can see also your stick movements again in the Blackbox Explorer.

image

Note that Throttle still isn't working properly - as before. @rotorflight Maybe you can look into this?

Test log with rcCommand[4] (collective) added.  

[LOG221.TXT](https://github.com/rotorflight/rotorflight-firmware/files/11129645/LOG221.TXT)

### CODE REVIEW

**Review by rotorflight — CHANGES_REQUESTED** [review]
Something is wrong with this change.
Collective was already in the log. It's rcCommand[3].  Throttle is rcCommand[4].
It is not logged. I guess you would like to add the missing throttle command?

**Comment by rotorflight** [comment]
https://github.com/rotorflight/rotorflight-firmware/blob/devel/src/main/fc/rc_controls.h#L34

File: src/main/blackbox/blackbox.c [hunk 1]
```diff
@@ -201,7 +201,8 @@ static const blackboxDeltaFieldDefinition_t blackboxMainFields[] =
     {"rcCommand",   0, SIGNED,   .Ipredict = PREDICT(0),       .Iencode = ENCODING(SIGNED_VB),    .Ppredict = PREDICT(PREVIOUS),      .Pencode = ENCODING(TAG8_4S16),  CONDITION(COMMAND)},
     {"rcCommand",   1, SIGNED,   .Ipredict = PREDICT(0),       .Iencode = ENCODING(SIGNED_VB),    .Ppredict = PREDICT(PREVIOUS),      .Pencode = ENCODING(TAG8_4S16),  CONDITION(COMMAND)},
     {"rcCommand",   2, SIGNED,   .Ipredict = PREDICT(0),       .Iencode = ENCODING(SIGNED_VB),    .Ppredict = PREDICT(PREVIOUS),      .Pencode = ENCODING(TAG8_4S16),  CONDITION(COMMAND)},
-    {"rcCommand",   3, SIGNED,   .Ipredict = PREDICT(0),       .Iencode = ENCODING(SIGNED_VB),    .Ppredict = PREDICT(PREVIOUS),      .Pencode = ENCODING(TAG8_4S16),  CONDITION(COMMAND)},
+    {"rcCommand",   3, UNSIGNED, .Ipredict = PREDICT(0),       .Iencode = ENCODING(UNSIGNED_VB),  .Ppredict = PREDICT(PREVIOUS),      .Pencode = ENCODING(TAG8_4S16), CONDITION(COMMAND)},
```
Comment (rotorflight) [inline src/main/blackbox/blackbox.c:204]: This line is incorrect. rcCommand[3] is collective, so it must be SIGNED

File: src/main/blackbox/blackbox.c [hunk 2]
```diff
@@ -201,7 +201,8 @@ static const blackboxDeltaFieldDefinition_t blackboxMainFields[] =
...
+    {"rcCommand",   4, SIGNED,   .Ipredict = PREDICT(0),       .Iencode = ENCODING(SIGNED_VB),    .Ppredict = PREDICT(PREVIOUS),      .Pencode = ENCODING(SIGNED_VB), CONDITION(COMMAND)},
```
Comment (rotorflight) [inline src/main/blackbox/blackbox.c:205]: And this is throttle, so it should be UNSIGNED

File: src/main/blackbox/blackbox.c [hunk 3]
```diff
@@ -641,7 +642,17 @@ static void writeIntraframe(void)
     blackboxWriteUnsignedVB(blackboxCurrent->time);
 
     if (testBlackboxCondition(CONDITION(COMMAND))) {
-        blackboxWriteSigned16VBArray(blackboxCurrent->command, 4);
+        // Write roll, pitch and yaw first:
+        blackboxWriteSigned16VBArray(blackboxCurrent->command, 3);
```
Comment (rotorflight) [inline src/main/blackbox/blackbox.c:646]: This is incorrect. Above it is defined that there are 4 fields of type TAG8_4S16. But here we are writing only 3.

File: src/main/blackbox/blackbox.c [hunk 4]
```diff
@@ -641,7 +642,17 @@ static void writeIntraframe(void)
...
+
+        /*
+        * Write the throttle separately from the rest of the RC data as it's unsigned.
+        * Throttle lies in range [PWM_RANGE_MIN..PWM_RANGE_MAX]:
+        */
+        blackboxWriteUnsignedVB(blackboxCurrent->command[THROTTLE]);
```
Comment (rotorflight) [inline src/main/blackbox/blackbox.c:652]: This is incorrect, for the same reason as above.

File: src/main/blackbox/blackbox.c [hunk 5]
```diff
@@ -773,6 +784,10 @@ static void writeInterframe(void)
     if (testBlackboxCondition(CONDITION(COMMAND))) {
         CALC_DELTAS(deltas, blackboxCurrent->command, blackboxPrev->command, 4);
         blackboxWriteTag8_4S16(deltas);
+
+        // Calculate collective delta
+        int32_t collectiveDelta = blackboxCurrent->command[COLLECTIVE] - blackboxPrev->command[COLLECTIVE];
```
Comment (rotorflight) [inline src/main/blackbox/blackbox.c:789]: COLLECTIVE is already handled by the code two line before

File: src/main/blackbox/blackbox.c (hunk 1 above)
Comment (egonl) [reply src/main/blackbox/blackbox.c:204]: Is it? Isn't it throttle? See the blackbox code.

File: src/main/blackbox/blackbox.c (hunk 2 above)
Comment (egonl) [reply src/main/blackbox/blackbox.c:205]: I think it's collective, see the blackbox explorer code.

**Pull request closed without merging**